elyra/pipeline/component.py

```diff
@@ -0,0 +1,297 @@
+        self._op = op
+        self._extension = extension
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def runtime(self):
+        return self._runtime
+
+    @property
+    def properties(self):
+        return self._properties
+
+    @property
+    def op(self):
+        if self._op:
+            return self._op
+        else:
+            return self._id
+
+    @property
+    def extension(self):
+        return self._extension
+
+
+class ComponentReader(LoggingConfigurable):
+    """
+    Abstract class to model component_entry readers that can read components from different locations
+    """
+    type: str = None
+
+    @property
+    def type(self) -> str:
+        return self.type
+
+    @abstractmethod
+    def read_component_definition(self, component_id: str, location: str) -> str:
+        raise NotImplementedError()
+
+
+class FilesystemComponentReader(ComponentReader):
+    """
+    Read a component_id definition from the local filesystem
+    """
+    type = 'filesystem'
+
+    def read_component_definition(self, component_id: str, location: str) -> str:
+        component_location = os.path.join(os.path.dirname(__file__), location)
+        if not os.path.exists(component_location):
+            self.log.error(f'Invalid location for component: {component_id} -> {component_location}')
+            raise FileNotFoundError(f'Invalid location for component: {component_id} -> {component_location}')
+
+        with open(component_location, 'r') as f:
+            return f.read()
+
+
+class UrlComponentReader(ComponentReader):
+    """
+    Read a component_id definition from a url
+    """
+    type = 'url'
+
+    def read_component_definition(self, component_id: str, location: str) -> str:
+        res = requests.get(location)
+        if res.status_code != HTTPStatus.OK:
+            self.log.error(f'Invalid location for component: {component_id} -> {location}')
+            raise FileNotFoundError(f'Invalid location for component: {component_id} -> {location}')
+
+        return res.text
+
+
+class ComponentParser(LoggingConfigurable):  # ABC
+    _readers = {
+        FilesystemComponentReader.type: FilesystemComponentReader(),
+        UrlComponentReader.type: UrlComponentReader()
+    }
+
+    @abstractmethod
+    def parse(self, registry_entry: dict) -> List[Component]:
+        raise NotImplementedError()
+
+    def _get_reader(self, component_entry):
+        """
+        Find the proper reader based on the given registry component_id.
+        """
+        if not component_entry:
+            raise ValueError("Invalid null component_id")
+
+        try:
+            return self._readers.get(component_entry.type)
+        except Exception:
+            raise ValueError(f'Unsupported registry type {component_entry.type}.')
```

Comment:
Should this be `reader type`?

Reply:
No, it defines the type of the catalog entry, it has no knowledge of there is a reader to read it.